source/fe/fe_p1nc.cc

```diff
@@ -0,0 +1,418 @@
+
+  c[0] = 0.25 - cpt(0)*a[0] - cpt(1)*b[0] ;
+  c[1] = 0.25 - cpt(0)*a[1] - cpt(1)*b[1] ;
+  c[2] = 0.25 - cpt(0)*a[2] - cpt(1)*b[2] ;
+  c[3] = 0.25 - cpt(0)*a[3] - cpt(1)*b[3] ;
+}
+
+
+
+FiniteElement<2,2>::InternalDataBase *
+FE_P1NC::get_data (const UpdateFlags update_flags,
+                   const Mapping<2,2> &,
+                   const Quadrature<2> &,
+                   dealii::internal::FEValues::FiniteElementRelatedData<2,2> &) const
+{
+  FiniteElement<2,2>::InternalDataBase *data = new FiniteElement<2,2>::InternalDataBase;
+
+  data->update_each = requires_update_flags(update_flags);
+
+  return data;
+}
+
+
+
+void
+FE_P1NC::fill_fe_values (const Triangulation<2,2>::cell_iterator           &cell,
+                         const CellSimilarity::Similarity                   ,
+                         const Quadrature<2>                               &quadrature,
+                         const Mapping<2,2>                                &mapping,
+                         const Mapping<2,2>::InternalDataBase &,
+                         const internal::FEValues::MappingRelatedData<2,2> &mapping_data,
+                         const FiniteElement<2,2>::InternalDataBase        &fe_internal,
+                         internal::FEValues::FiniteElementRelatedData<2,2> &output_data) const
+{
+  const UpdateFlags flags(fe_internal.update_each) ;
+
+  const unsigned int n_q_points = mapping_data.quadrature_points.size();
+
+  std::vector<double> values(flags & update_values ? this->dofs_per_cell : 0);
+  std::vector<Tensor<1,2> > grads(flags & update_gradients ? this->dofs_per_cell : 0);
+
+
+  // linear shape with a half value: phi(x,y) = ax + by + c
+  std::vector<double> a(4), b(4), c(4);
+  get_linear_shape (cell, a, b, c);
+
+
+  // compute basis functions
+  if (flags & (update_values | update_gradients))
+    for (unsigned int i=0; i<n_q_points; ++i)
+      {
+        for (unsigned int k=0; k<this->dofs_per_cell; ++k)
+          {
+            if (flags & update_values)
+              {
+                values[k] = a[k]*mapping_data.quadrature_points[i](0) + b[k]*mapping_data.quadrature_points[i](1) + c[k] ;
+                output_data.shape_values[k][i] = values[k];
+              }
+
+            if (flags & update_gradients)
+              {
+                grads[k][0] = a[k] ;
+                grads[k][1] = b[k] ;
+                output_data.shape_gradients[k][i] = grads[k];
+              }
+          }
+      }
+
+  // hessian
+  std::vector<Tensor<2,2> > hessians(flags & update_hessians ? this->dofs_per_cell : 0);
+  if (flags & update_hessians)
+    {
+      for (unsigned int i=0; i<n_q_points; ++i)
+        {
+          for (unsigned int k=0; k<this->dofs_per_cell; ++k)
+            {
+              hessians[k][0][0] = 0.0 ;
+              hessians[k][0][1] = 0.0 ;
+              hessians[k][1][0] = 0.0 ;
+              hessians[k][1][1] = 0.0 ;
+              output_data.shape_hessians[k][i] = hessians[k];
```

Comment:
This whole block can be written more efficiently if you know that `Tensor<2,2>()` creates a tensor of all zeros. Then, you should be able to do
```
if (flags & update_hessians)
  std::fill (output_data.shape_hessians.begin(),
               output_data.shape_hessians.end(),
              Tensor<2,2>());
```
In fact, since this would be the same every time, you may already want to do exactly this in `get_data()`, `get_face_data()`, and `get_subface_data()` and then delete the `update_hessians` case from this function altogether.